### DISCUSSION TITLE
Hardware supporting PTP/PPS with IEEE1588-2008 on the Compute Module 4

### DISCUSSION THREAD

**Discussion by geerlingguy** [geerlingguy/raspberry-pi-pcie-devices#457] [General]
I've opened this discussion to allow for more focused discussion on hardware implementations (HATs, PCIe devices, etc.) that can be used with the Compute Module 4 to enhance it's IEEE1588-2008 capabilities, now that [this issue](https://github.com/raspberrypi/linux/issues/4151) seems to be resolved (or nearly so).

**Comment by ddivins** [comment]
All,  I just wanted to take a moment to give a big shout out to Lasse and Ian from http://timebeat.app

I have my CM4 with the latest kernel doing HW timestamping PTP with Timebeat Enterprise and NTP being servered by Chrony (in free mode -x).  They also walked me through some tweaks on of the container based elastic search/grafana.

With an Uptronics hat (manually jumpered to sync_in) I am +/- 20ns.

Look forward to seeing what these new hats can do!

**Reply by 652436962** [reply]
> All, I just wanted to take a moment to give a big shout out to Lasse and Ian from http://timebeat.app
> 
> I have my CM4 with the latest kernel doing HW timestamping PTP with Timebeat Enterprise and NTP being servered by Chrony (in free mode -x). They also walked me through some tweaks on of the container based elastic search/grafana.
> 
> With an Uptronics hat (manually jumpered to sync_in) I am +/- 20ns.
> 
> Look forward to seeing what these new hats can do!

hello  can you provide your img ,  i can't  rpi-update

**Comment by lasselj** [comment]
The first batch of breakout boards have arrived. We have tested and are shipping today to those that ordered on https://store.timebeat.app/

I can't wait to hear what people do with them :-)

![cm4_bob_1](https://user-images.githubusercontent.com/12864656/178006635-05d46934-108b-45c5-b66a-b2bb9fc78284.png)
<img width="1200" alt="cm4_bob_4" src="https://user-images.githubusercontent.com/12864656/178006660-2e9644fc-9c1e-41e4-be15-632110462810.png">
![cm4_bob_2](https://user-images.githubusercontent.com/12864656/178006663-9493f687-94cc-42a2-8335-5ad9c770c83a.png)
![cm4_bob_3](https://user-images.githubusercontent.com/12864656/178006669-4e0073ec-0786-452e-8ab8-ac9dabdfccf4.png)

**Reply by bmaya** [reply]
I like the uFL adapter - and thinking the IOBoard connector should have been that to begin with for the SYNC lines

**Reply by lasselj** [reply]
In that case, you'll appreciate that [this product](https://store.timebeat.app/products/raspberry-pi-cm4-1pps-breakout-board) now comes in ENIG gold and black instead of green and white.... I don't know why, but somehow I've decided we are on a mission to make everything look sort of like this.... :-)

![John Player Special](https://user-images.githubusercontent.com/12864656/221444756-f85d9039-5a05-4d70-92bb-b29fc79460a2.jpg)

**Comment by bmaya** [comment]
I've been working a GPS clock hat and was really excited by the PTP potential. Turns out it works best on the CM4 rather than the pi4 but oh well.

I was aiming for an anything with a 40pin pi header GPS clock source.

I also have a clock sync and de-jitter circuit on the board for external reference I/O pulses.

With some tuning and hacking around chrony v0.1 still gets the system clock ~100ns from UTC, and RMS offset of around the same. Probing here and there and it turns out it's all good :-)
![clockHatV0 1](https://user-images.githubusercontent.com/3286975/178461533-e1d600cd-1745-49e9-b9ea-806f02c86d31.jpeg)

Next versions are tidied up, with and without the sync/jitter circuit. And u.FL connectors for SYNC. They're a dummy check and GND stitchiing away from being ordered. 
![Screenshot from 2022-07-12 10-31-54](https://user-images.githubusercontent.com/3286975/178462339-13871236-cfc0-4afb-a3c3-131bec86920c.png)
![Screenshot from 2022-07-12 10-28-47](https://user-images.githubusercontent.com/3286975/178462375-6966d5bc-ade0-4ca6-bf6a-f6a57ef6cb29.png)

(no idea what happened to my 3d models but right now not fighting whatever made them disappear)

Graphs of time/clock performance coming soon.

**Comment by lasselj** [comment]
We made this "sandwich" board.... they work a treat (different options from £45). We wanted to avoid hawking the 40 pin connector so it can be used for other purposes. Only the CM4 supports hardware timestamping currently and you can't get the pulse to the right place using the 40 pin connector. So anyhow, here it is... :

![Lea_5f569772-f7cc-452f-8bd4-76dcf41bb95c](https://user-images.githubusercontent.com/12864656/200043119-b97738ca-2c0b-4960-8c4c-3e1bdafc0bf5.png)

<img width="680" alt="base-cm4" src="https://user-images.githubusercontent.com/12864656/200043132-ffc9e2b1-b3d6-4599-a8db-c1118d23496c.png">


More info here: https://store.timebeat.app/products/gnss-raspberry-pi-cm4-module

All the best,

Lasse

**Reply by lasselj** [reply]
In response to myself, I will mention that we now also have some fairly cool OCXO modules from SiTime that goes in the expansion slots and whereas they are not exactly low cost, they do provide 5ppb which (with the greatest respect) is spectacularly awesome on a CM4 and if you buy them with a Timecard Mini we extend a rather large discount. (URL still the same as above and they are currently "in stock" - which is rare :-))
